presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/GroupInnerJoinsByConnector.java

```diff
@@ -612,136 +586,87 @@ public boolean getContainsCombinedSources()
-                        return;
+                if (!(resolved instanceof JoinNode)) {
+                    if (resolved instanceof ProjectNode) {
+                        /*
+                            Certain ProjectNodes can be 'inlined' into the parent TableScan, e.g a CAST expression
+                            We will do this here while flattening the JoinNode if possible
+                            For now, we log the fact that we saw a ProjectNode and if identity projection, will continue
+                        */
+                        //Only identity projections can be handled.
+                        if (AssignmentUtils.isIdentity(((ProjectNode) resolved).getAssignments())) {
+                            flattenNode(((ProjectNode) resolved).getSource());
+                            return;
+                        }
                     }
+                    sources.add(resolved);
+                    return;
+                }
+                JoinNode joinNode = (JoinNode) resolved;
+                if (joinNode.getType() != INNER || !determinismEvaluator.isDeterministic(joinNode.getFilter().orElse(TRUE_CONSTANT))) {
+                    sources.add(resolved);
+                    return;
                 }
-                sources.add(resolved);
-                return;
-            }
-            JoinNode joinNode = (JoinNode) resolved;
-            if (joinNode.getType() != INNER || !determinismEvaluator.isDeterministic(joinNode.getFilter().orElse(TRUE_CONSTANT))) {
-                sources.add(resolved);
-                return;
-            }
-
-            flattenNode(joinNode.getLeft());
-            flattenNode(joinNode.getRight());
-            joinNode.getCriteria().stream()
-                    .map(criteria -> toRowExpression(criteria, functionResolution))
-                    .forEach(joinCriteriaFilters::add);
-            joinNode.getFilter().ifPresent(joinCriteriaFilters::add);
-        }
-
-        MultiJoinNode toMultiJoinNode()
-        {
-            ImmutableSet<VariableReferenceExpression> inputVariables = sources.stream().flatMap(source -> source.getOutputVariables().stream()).collect(toImmutableSet());
 
-            /*
-             We do this to satisfy the invariant that the rewritten Join node must produce the same output variables as the input Join node
-            */
+                flattenNode(joinNode.getLeft());
+                flattenNode(joinNode.getRight());
+                joinNode.getCriteria().stream()
+                        .map(criteria -> toRowExpression(criteria, functionResolution))
+                        .forEach(joinCriteriaFilters::add);
+                joinNode.getFilter().ifPresent(joinCriteriaFilters::add);
+            }
 
-            ImmutableSet.Builder<VariableReferenceExpression> updatedOutputVariables = ImmutableSet.builder();
+            GroupInnerJoinsMultiJoinNode toMultiJoinNode()
```

Comment:
Similarly, this class can have its own `buildMultiJoinNode` with its own logic

Reply:
I didn't understand what is better in doing like this, or how to do it.

Can you check the [commit](https://github.com/Haritha-Koloth/presto/pull/1/commits/8de7fc4b3bcb837e403728b64154c04a4d2abefb) and see if this is what u meant ?